# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/BytecodeUtil.cpp
# Commit: 5f6b9e2fc3bb
# Full Hash: 5f6b9e2fc3bb3c5fd74f9871399269a2665a5270
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Refactor try/finally to use goto r=jandem
#   
#   JSOp::Gosub is just a JSOp::Goto that lies about its effect on the stack. We can start simplifying try/finally by using a JSOp::Goto instead.
#   
#   Currently, to make the stack depth work out when returning from a finally, Gosub pretends to pop two values (the arguments to retsub), and Finally pretends to push them. This means that bytecode analysis can treat Gosub as if it falls through, and the stack depth of the jumptarget following the gosub will be correct. This patch uses a different approach: bytecode analysis marks the return address as reachable when we push it using JSOp::ResumeIndex. This means that the bytecode no longer has to lie about stack depth.
# ==============================================================================

diff -r b199b21ed5a3 -r 5f6b9e2fc3bb js/src/vm/BytecodeUtil.cpp
--- a/js/src/vm/BytecodeUtil.cpp	Tue Mar 08 21:00:30 2022 +0000
+++ b/js/src/vm/BytecodeUtil.cpp	Tue Mar 08 21:10:21 2022 +0000
@@ -926,7 +926,13 @@
                 return false;
               }
             } else if (tn.kind() == TryNoteKind::Finally) {
-              if (!addJump(catchOffset, stackDepth, offsetStack, pc,
+              // Two additional values will be on the stack at the beginning
+              // of the finally block: the exception/resume index, and the
+              // |throwing| value. For the benefit of the decompiler, point
+              // them at this Try.
+              offsetStack[stackDepth].set(offset, 0);
+              offsetStack[stackDepth + 1].set(offset, 1);
+              if (!addJump(catchOffset, stackDepth + 2, offsetStack, pc,
                            JumpKind::TryFinally)) {
                 return false;
               }
@@ -936,6 +942,18 @@
         break;
       }
 
+      case JSOp::ResumeIndex: {
+        // ResumeIndex is used to push a return address for a finally block. If
+        // this op is reachable, then so is that return address (with a smaller
+        // stack depth because the resume index and `throwing` will have been
+        // popped.
+        uint32_t resumeOffset = script_->resumeOffsets()[GET_UINT24(pc)];
+        if (!recordBytecode(resumeOffset, offsetStack, stackDepth - 2)) {
+          return false;
+        }
+        break;
+      }
+
       default:
         break;
     }
@@ -2094,7 +2112,9 @@
       case JSOp::Exception:
         return write("EXCEPTION");
 
-      case JSOp::Finally:
+      case JSOp::Try:
+        // Used for the values live on entry to the finally block.
+        // See TryNoteKind::Finally above.
         if (defIndex == 0) {
           return write("THROWING");
         }